roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -0,0 +1,247 @@
+    // leaving. Sorted ascending by core index.
+    candidates: Vec<(CoreIndex, Hash)>,
+    // A bitfield where the i'th bit corresponds to the i'th candidate in `candidates`.
+    // The i'th bit is `true` iff the candidate has been approved in the context of
+    // this block. The block can be considered approved has all bits set to 1
+    approved_bitfield: Bitfield,
+    rotation_offset: GroupIndex,
+    children: Vec<Hash>,
+}
+
+// slot_duration * 2 + DelayTranche gives the number of delay tranches since the
+// unix epoch.
+type Tick = u64;
+
+struct TrackerEntry 
+
+struct StoredBlockRange(BlockNumber, BlockNumber)
+```
+
+In the schema, we map
+
+```
+"StoredBlocks" => StoredBlockRange
+BlockNumber => Vec<BlockHash>
+BlockHash => BlockEntry
+CandidateHash => CandidateEntry
+```
+
+## Logic
+
+In-memory state:
+
+```rust
+struct ApprovalVoteRequest {
+  validator_index: ValidatorIndex,
+  block_hash: Hash,
+  candidate_index: u32,
+}
+
+struct State {
+    earliest_session: SessionIndex,
+    session_info: Vec<SessionInfo>,
+    keystore: KeyStorePtr,
+    wakeups: BTreeMap<Tick, Vec<(Hash, Hash)>>, // Tick -> [(Relay Block, Candidate Hash)]
+    
+    // These are connected to each other.
+    approval_vote_tx: mpsc::Sender<ApprovalVoteRequest>,
+    approval_vote_rx: mpsc::Receiver<ApprovalVoteRequest>,
+}
+```
+
+[`SessionInfo`](../../runtime/session_info.md)
+
+On start-up, we clear everything currently stored by the database. This is done by loading the `StoredBlockRange`, iterating through each block number, iterating through each block hash, and iterating through each candidate referenced by each block. Although this is `O(o*n*p)`, we don't expect to have more than a few unfinalized blocks at any time and in extreme cases, a few thousand. The clearing operation should be relatively fast as a result.
+
+Main loop:
+  * Each iteration, select over all of
+    * The next `Tick` in `wakeups`: trigger `wakeup_process` for each `(Hash, Hash)` pair scheduled under the `Tick` and then remove all entries under the `Tick`.
+    * The next message from the overseer: handle the message as described in the [Incoming Messages section](#incoming-messages)
+    * The next request from `approval_vote_rx`: handle with `issue_approval`
+
+### Incoming Messages
+
+#### `OverseerSignal::BlockFinalized`
+
+On receiving an `OverseerSignal::BlockFinalized(h)`, we fetch the block number `b` of that block from the ChainApi subsystem. We update our `StoredBlockRange` to begin at `b+1`. Additionally, we remove all block entries and candidates referenced by them up to and including `b`. Lastly, we prune out all descendents of `h` transitively: when we remove a `BlockEntry` with number `b` that is not equal to `h`, we recursively delete all the `BlockEntry`s referenced as children. We remove the `block_assignments` entry for the block hash and if `block_assignments` is now empty, remove the `CandidateEntry`.
```

Comment:
We just need to ensure anything we want remembered makes it on-chain by this point, but I think that's done here somewhere.

Reply:
haven't done the on-chain reporting stuff yet. Disputes will track a DB of all votes we've seen by validators.

Reply:
I'd agree disputes provides a better venue for tracking this, thanks.